File: agent_lifecycle_protocol/registry.py
"""Fork registry and lineage tracking (Section 6).

Provides genealogical queries: ancestors, descendants, siblings,
family tree, genetic match, epigenetic match.
Implements access control (public/authorized/operator_only) and
decommissioned agent redaction.
"""

import hashlib
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Set

from .schema import (
    ACCESS_LEVELS,
    REDACTION_LEVELS,
    AgentRecord,
)
from .store import LifecycleStore


@dataclass
class RegistryEntry:
    """A registry entry for lineage queries (Section 6.2)."""
    agent_id: str = ""
    parent_id: Optional[str] = None
    children: List[str] = field(default_factory=list)
    siblings: List[str] = field(default_factory=list)
    generation: int = 1
    fork_type: str = ""
    genesis_timestamp: str = ""
    lifecycle_status: str = "provisioning"
    genetic_profile: Dict[str, Any] = field(default_factory=dict)
    epigenetic_profile: Dict[str, Any] = field(default_factory=dict)
    redaction_level: str = "none"
    last_updated: str = ""

    def to_dict(self) -> Dict[str, Any]:
        return {
            "agent_id": self.agent_id,
            "parent_id": self.parent_id,
            "children": list(self.children),
            "siblings": list(self.siblings),
            "generation": self.generation,
            "fork_type": self.fork_type,
            "genesis_timestamp": self.genesis_timestamp,
            "lifecycle_status": self.lifecycle_status,
            "genetic_profile": dict(self.genetic_profile),
            "epigenetic_profile": dict(self.epigenetic_profile),
            "redaction_level": self.redaction_level,
            "last_updated": self.last_updated,
        }

    @classmethod
    def from_agent(cls, agent: AgentRecord) -> "RegistryEntry":
        return cls(
            agent_id=agent.agent_id,
            parent_id=agent.parent_id,
            children=list(agent.children),
            generation=agent.generation,
            lifecycle_status=agent.state,
            genetic_profile=agent.genetic_profile.to_dict(),
            epigenetic_profile=agent.epigenetic_profile.to_dict(),
            redaction_level=agent.redaction_level,
            last_updated=agent.updated_at,
            genesis_timestamp=agent.created_at,
        )


def _pseudonymous_hash(agent_id: str) -> str:
    """Generate pseudonymous hash for redacted entries (Section 6.5)."""
    return "anon:" + hashlib.sha256(agent_id.encode("utf-8")).hexdigest()[:16]


class LineageRegistry:
    """Fork registry with lineage queries (Section 6.3).

    Builds an in-memory graph from the LifecycleStore's agent records.
    """

    def __init__(self, store: LifecycleStore) -> None:
        self.store = store

    def _build_index(self) -> Dict[str, AgentRecord]:
        """Build latest-snapshot index of all agents."""
        latest: Dict[str, AgentRecord] = {}
        for a in self.store.get_all_agents():
            latest[a.agent_id] = a
        return latest

    def get_entry(
        self, agent_id: str, access_level: str = "public"
    ) -> Optional[RegistryEntry]:
        """Get a registry entry with access-level filtering (Section 6.5)."""
        index = self._build_index()
        agent = index.get(agent_id)
        if agent is None:
            return None

        entry = RegistryEntry.from_agent(agent)

        # Compute siblings
        if agent.parent_id and agent.parent_id in index:
            parent = index[agent.parent_id]
            entry.siblings = [
                c for c in parent.children if c != agent_id
            ]

        # Apply redaction for decommissioned agents
        entry = self._apply_redaction(entry)

        # Apply access level filtering
        entry = self._apply_access_filter(entry, access_level)

        return entry

    def ancestors(self, agent_id: str, max_depth: int = 100) -> List[str]:
        """Return the complete ancestor chain (Section 6.3)."""
        index = self._build_index()
        chain: List[str] = []
        current = agent_id
        depth = 0
        while depth < max_depth:
            agent = index.get(current)
            if agent is None or agent.parent_id is None:
                break
            chain.append(agent.parent_id)
            current = agent.parent_id
            depth += 1
        return chain

    def descendants(
        self, agent_id: str, max_depth: int = 100
    ) -> List[str]:
        """Return all agents forked from this agent, recursively."""
        index = self._build_index()
        result: List[str] = []
        queue: List[str] = [agent_id]
        visited: Set[str] = {agent_id}
        depth = 0

        while queue and depth < max_depth:
            next_queue: List[str] = []
            for aid in queue:
                agent = index.get(aid)
                if agent is None:
                    continue
                for child in agent.children:
                    if child not in visited:
                        visited.add(child)
                        result.append(child)
                        next_queue.append(child)
            queue = next_queue
            depth += 1

        return result

    def siblings(self, agent_id: str) -> List[str]:
        """Return all agents sharing the same parent (Section 6.3)."""
        index = self._build_index()
        agent = index.get(agent_id)
        if agent is None or agent.parent_id is None:
            return []

        parent = index.get(agent.parent_id)
        if parent is None:
            return []

        return [c for c in parent.children if c != agent_id]

    def family_tree(
        self, agent_id: str, max_depth: int = 100
    ) -> Dict[str, Any]:
        """Return the complete genealogy tree (Section 6.3)."""
        index = self._build_index()

        def _build_node(aid: str, depth: int) -> Dict[str, Any]:
            agent = index.get(aid)
            if agent is None:
                return {"agent_id": aid, "status": "unknown", "children": []}
            node: Dict[str, Any] = {
                "agent_id": aid,
                "state": agent.state,
                "generation": agent.generation,
                "children": [],
            }
            if depth < max_depth:
                for child_id in agent.children:
                    node["children"].append(_build_node(child_id, depth + 1))
            return node

        # Walk up to root
        root = agent_id
        while True:
            agent = index.get(root)
            if agent is None or agent.parent_id is None:
                break
            root = agent.parent_id

        return _build_node(root, 0)

    def genetic_match(
        self, model_family: str = "", architecture: str = ""
    ) -> List[str]:
        """Find agents sharing genetic lineage (Section 6.3)."""
        index = self._build_index()
        results: List[str] = []
        for agent in index.values():
            if model_family and agent.genetic_profile.model_family != model_family:
                continue
            if architecture and agent.genetic_profile.architecture != architecture:
                continue
            results.append(agent.agent_id)
        return results

    def epigenetic_match(
        self, role: str = "", specialization: str = ""
    ) -> List[str]:
        """Find agents sharing epigenetic profiles (Section 6.3)."""
        index = self._build_index()
        results: List[str] = []
        for agent in index.values():
            if role and agent.epigenetic_profile.role != role:
                continue
            if specialization and agent.epigenetic_profile.specialization != specialization:
                continue
            results.append(agent.agent_id)
        return results

    def redact_entry(
        self, agent_id: str, level: str = "partial"
    ) -> Optional[AgentRecord]:
        """Apply redaction to a decommissioned agent's registry entry (Section 9.5)."""
        if level not in REDACTION_LEVELS:
            raise ValueError(f"Invalid redaction level: {level}")

        agent = self.store.get_agent(agent_id)
        if agent is None:
            return None
        if agent.state != "decommissioned":
            raise ValueError("Only decommissioned agents can be redacted")

        agent.redaction_level = level
        self.store.save_agent(agent)
        return agent

    def _apply_redaction(self, entry: RegistryEntry) -> RegistryEntry:
        """Apply redaction based on the entry's redaction level."""
        if entry.redaction_level == "none":
            return entry

        if entry.redaction_level == "partial":
            # Keep: agent_id, lineage links, genetic_profile, status, timestamp
            # Remove: epigenetic_profile details
            entry.epigenetic_profile = {}
            return entry

        if entry.redaction_level == "full":
            # Replace agent_id with pseudonymous hash
            original_id = entry.agent_id
            entry.agent_id = _pseudonymous_hash(original_id)
            if entry.parent_id:
                entry.parent_id = _pseudonymous_hash(entry.parent_id)
            entry.children = [_pseudonymous_hash(c) for c in entry.children]
            entry.siblings = [_pseudonymous_hash(s) for s in entry.siblings]
            entry.genetic_profile = {}
            entry.epigenetic_profile = {}
            return entry

        return entry

    def _apply_access_filter(
        self, entry: RegistryEntry, access_level: str
    ) -> RegistryEntry:
        """Filter entry fields based on access level (Section 6.5)."""
        if access_level == "operator_only":
            return entry  # full access

        if access_level == "authorized":
            # Remove operator-only fields
            entry.epigenetic_profile = {}
            return entry

        # Public access: minimal fields
        entry.epigenetic_profile = {}
        entry.children = []
        entry.siblings = []
        return entry

File: typescript/src/registry.ts
import { createHash } from "node:crypto";
import {
  REDACTION_LEVELS,
  AgentRecord,
} from "./types";
import { LifecycleStore } from "./store";

export interface RegistryEntryData {
  agent_id: string;
  parent_id: string | null;
  children: string[];
  siblings: string[];
  generation: number;
  fork_type: string;
  genesis_timestamp: string;
  lifecycle_status: string;
  genetic_profile: Record<string, unknown>;
  epigenetic_profile: Record<string, unknown>;
  redaction_level: string;
  last_updated: string;
}

export class RegistryEntry {
  agentId: string;
  parentId: string | null;
  children: string[];
  siblings: string[];
  generation: number;
  forkType: string;
  genesisTimestamp: string;
  lifecycleStatus: string;
  geneticProfile: Record<string, unknown>;
  epigeneticProfile: Record<string, unknown>;
  redactionLevel: string;
  lastUpdated: string;

  constructor(opts?: Partial<{
    agentId: string;
    parentId: string | null;
    children: string[];
    siblings: string[];
    generation: number;
    forkType: string;
    genesisTimestamp: string;
    lifecycleStatus: string;
    geneticProfile: Record<string, unknown>;
    epigeneticProfile: Record<string, unknown>;
    redactionLevel: string;
    lastUpdated: string;
  }>) {
    this.agentId = opts?.agentId ?? "";
    this.parentId = opts?.parentId ?? null;
    this.children = opts?.children ?? [];
    this.siblings = opts?.siblings ?? [];
    this.generation = opts?.generation ?? 1;
    this.forkType = opts?.forkType ?? "";
    this.genesisTimestamp = opts?.genesisTimestamp ?? "";
    this.lifecycleStatus = opts?.lifecycleStatus ?? "provisioning";
    this.geneticProfile = opts?.geneticProfile ?? {};
    this.epigeneticProfile = opts?.epigeneticProfile ?? {};
    this.redactionLevel = opts?.redactionLevel ?? "none";
    this.lastUpdated = opts?.lastUpdated ?? "";
  }

  toDict(): RegistryEntryData {
    return {
      agent_id: this.agentId,
      parent_id: this.parentId,
      children: [...this.children],
      siblings: [...this.siblings],
      generation: this.generation,
      fork_type: this.forkType,
      genesis_timestamp: this.genesisTimestamp,
      lifecycle_status: this.lifecycleStatus,
      genetic_profile: { ...this.geneticProfile },
      epigenetic_profile: { ...this.epigeneticProfile },
      redaction_level: this.redactionLevel,
      last_updated: this.lastUpdated,
    };
  }

  static fromAgent(agent: AgentRecord): RegistryEntry {
    return new RegistryEntry({
      agentId: agent.agentId,
      parentId: agent.parentId,
      children: [...agent.children],
      generation: agent.generation,
      lifecycleStatus: agent.state,
      geneticProfile: { ...agent.geneticProfile.toDict() },
      epigeneticProfile: { ...agent.epigeneticProfile.toDict() },
      redactionLevel: agent.redactionLevel,
      lastUpdated: agent.updatedAt,
      genesisTimestamp: agent.createdAt,
    });
  }
}

function pseudonymousHash(agentId: string): string {
  return (
    "anon:" +
    createHash("sha256").update(agentId, "utf-8").digest("hex").slice(0, 16)
  );
}

export class LineageRegistry {
  readonly store: LifecycleStore;

  constructor(store: LifecycleStore) {
    this.store = store;
  }

  private buildIndex(): Map<string, AgentRecord> {
    const latest = new Map<string, AgentRecord>();
    for (const a of this.store.getAllAgents()) {
      latest.set(a.agentId, a);
    }
    return latest;
  }

  getEntry(
    agentId: string,
    accessLevel: string = "public",
  ): RegistryEntry | null {
    const index = this.buildIndex();
    const agent = index.get(agentId);
    if (!agent) return null;

    let entry = RegistryEntry.fromAgent(agent);

    if (agent.parentId && index.has(agent.parentId)) {
      const parent = index.get(agent.parentId)!;
      entry.siblings = parent.children.filter((c) => c !== agentId);
    }

    entry = this.applyRedaction(entry);
    entry = this.applyAccessFilter(entry, accessLevel);

    return entry;
  }

  ancestors(agentId: string, maxDepth: number = 100): string[] {
    const index = this.buildIndex();
    const chain: string[] = [];
    let current = agentId;
    let depth = 0;
    while (depth < maxDepth) {
      const agent = index.get(current);
      if (!agent || !agent.parentId) break;
      chain.push(agent.parentId);
      current = agent.parentId;
      depth++;
    }
    return chain;
  }

  descendants(agentId: string, maxDepth: number = 100): string[] {
    const index = this.buildIndex();
    const result: string[] = [];
    let queue = [agentId];
    const visited = new Set([agentId]);
    let depth = 0;

    while (queue.length > 0 && depth < maxDepth) {
      const nextQueue: string[] = [];
      for (const aid of queue) {
        const agent = index.get(aid);
        if (!agent) continue;
        for (const child of agent.children) {
          if (!visited.has(child)) {
            visited.add(child);
            result.push(child);
            nextQueue.push(child);
          }
        }
      }
      queue = nextQueue;
      depth++;
    }

    return result;
  }

  siblings(agentId: string): string[] {
    const index = this.buildIndex();
    const agent = index.get(agentId);
    if (!agent || !agent.parentId) return [];

    const parent = index.get(agent.parentId);
    if (!parent) return [];

    return parent.children.filter((c) => c !== agentId);
  }

  familyTree(
    agentId: string,
    maxDepth: number = 100,
  ): Record<string, unknown> {
    const index = this.buildIndex();

    const buildNode = (
      aid: string,
      depth: number,
    ): Record<string, unknown> => {
      const agent = index.get(aid);
      if (!agent) {
        return { agent_id: aid, status: "unknown", children: [] };
      }
      const node: Record<string, unknown> = {
        agent_id: aid,
        state: agent.state,
        generation: agent.generation,
        children: [] as Record<string, unknown>[],
      };
      if (depth < maxDepth) {
        (node.children as Record<string, unknown>[]) = agent.children.map(
          (childId) => buildNode(childId, depth + 1),
        );
      }
      return node;
    };

    let root = agentId;
    while (true) {
      const agent = index.get(root);
      if (!agent || !agent.parentId) break;
      root = agent.parentId;
    }

    return buildNode(root, 0);
  }

  geneticMatch(
    modelFamily: string = "",
    architecture: string = "",
  ): string[] {
    const index = this.buildIndex();
    const results: string[] = [];
    for (const agent of index.values()) {
      if (modelFamily && agent.geneticProfile.modelFamily !== modelFamily)
        continue;
      if (architecture && agent.geneticProfile.architecture !== architecture)
        continue;
      results.push(agent.agentId);
    }
    return results;
  }

  epigeneticMatch(
    role: string = "",
    specialization: string = "",
  ): string[] {
    const index = this.buildIndex();
    const results: string[] = [];
    for (const agent of index.values()) {
      if (role && agent.epigeneticProfile.role !== role) continue;
      if (
        specialization &&
        agent.epigeneticProfile.specialization !== specialization
      )
        continue;
      results.push(agent.agentId);
    }
    return results;
  }

  redactEntry(
    agentId: string,
    level: string = "partial",
  ): AgentRecord | null {
    if (!(REDACTION_LEVELS as readonly string[]).includes(level)) {
      throw new Error(`Invalid redaction level: ${level}`);
    }

    const agent = this.store.getAgent(agentId);
    if (agent === null) return null;
    if (agent.state !== "decommissioned") {
      throw new Error("Only decommissioned agents can be redacted");
    }

    agent.redactionLevel = level;
    this.store.saveAgent(agent);
    return agent;
  }

  private applyRedaction(entry: RegistryEntry): RegistryEntry {
    if (entry.redactionLevel === "none") return entry;

    if (entry.redactionLevel === "partial") {
      entry.epigeneticProfile = {};
      return entry;
    }

    if (entry.redactionLevel === "full") {
      const originalId = entry.agentId;
      entry.agentId = pseudonymousHash(originalId);
      if (entry.parentId) {
        entry.parentId = pseudonymousHash(entry.parentId);
      }
      entry.children = entry.children.map(pseudonymousHash);
      entry.siblings = entry.siblings.map(pseudonymousHash);
      entry.geneticProfile = {};
      entry.epigeneticProfile = {};
      return entry;
    }

    return entry;
  }

  private applyAccessFilter(
    entry: RegistryEntry,
    accessLevel: string,
  ): RegistryEntry {
    if (accessLevel === "operator_only") return entry;

    if (accessLevel === "authorized") {
      entry.epigeneticProfile = {};
      return entry;
    }

    entry.epigeneticProfile = {};
    entry.children = [];
    entry.siblings = [];
    return entry;
  }
}
